File: cli/src/freepod/delete.py
"""Deleting the deployment a project points at.

The mirror image of `deploy`, and deliberately the only destructive thing this
client does. It addresses the deployment recorded in `.freepod.json` and
nothing else: a command that could name an arbitrary deployment would be one
whose worst typo is unrecoverable, and the project file is the only place the
client knows a deployment by.

Teardown is asynchronous. `DELETE` answers **204** and moves the deployment to
`deleting`; the reconciler uninstalls the release and removes the namespace
afterwards, which is what actually destroys the volumes and frees the hostname.
The client therefore follows the teardown to `deleted` by default — not for
ceremony, but because the hostname stays claimed until it lands, so a `delete`
followed immediately by a `deploy` would otherwise collide with the deployment
it just removed.

The pointer is discarded from the project file the moment the platform accepts
the deletion, before the teardown is followed. From that instant the deployment
can never serve this project again — an update is refused for anything outside
`ready`/`error` — so keeping the pointer would only make the next `deploy`
fail on a deployment the user already asked to be rid of. The user values stay
put: the hostname is intent, and re-deploying should re-claim the same name.
"""

from __future__ import annotations

import sys
import time
from pathlib import Path
from typing import Any, Callable, Dict, Optional

import click

from . import FreepodError, UsageError
from .api import ApiClient, _json_detail
from .config import ROLLOUT_WAIT_SECONDS
from .deploy import describe_conflict
from .project import PROJECT_FILE, Project, require_project

#: Where a teardown ends. `deleted` is the record the platform keeps behind;
#: a 404 means it stopped keeping one, and both mean the same thing here.
STATUS_DELETED = "deleted"

#: In flight. Requesting a deletion again from here is accepted and changes
#: nothing, so it is a state to wait out rather than an error.
STATUS_DELETING = "deleting"

STATUS_ERROR = "error"

#: How often to re-read while following a teardown. Matched to the deploy
#: pipeline's cadence for the same reason: the reconciler's own work is
#: measured in tens of seconds.
POLL_SECONDS = 3.0


def _log(message: str) -> None:
    print(message, file=sys.stderr)


def _ask(question: str) -> bool:
    return click.confirm(question, default=False, err=True)


# --------------------------------------------------------------------------
# Reads
# --------------------------------------------------------------------------


def read_deployment(
    api: ApiClient, user_id: int, deployment_id: str
) -> Optional[Dict[str, Any]]:
    """The deployment, or None if the platform no longer has it.

    A 404 is an ordinary answer here rather than a failure: the pointer lives
    in a committed file and outliving the deployment is exactly what it does.
    """
    response = api.get(f"/api/users/{user_id}/deployments/{deployment_id}")
    if response.status_code == 404:
        return None
    if not response.is_success:
        raise FreepodError(
            f"could not read deployment {deployment_id}: HTTP {response.status_code} "
            f"{response.text.strip()[:300]}"
        )
    return response.json()


def address(deployment: Dict[str, Any]) -> Optional[str]:
    hostname = deployment.get("hostname")
    return f"https://{hostname}" if hostname else None


# --------------------------------------------------------------------------
# Asking
# --------------------------------------------------------------------------


def present(deployment: Dict[str, Any], env_name: str, echo: Callable[[str], None]) -> None:
    """Show what is about to be destroyed, by name and by address.

    The name alone is not enough to recognize a deployment — they are generated
    and interchangeable at a glance — so the address is shown too. It is the
    part a user actually knows.
    """
    echo("")
    echo(f"About to delete deployment '{deployment.get('name')}' on '{env_name}':")
    live = address(deployment)
    if live:
        echo(f"  address  {live}")
    echo(f"  status   {deployment.get('status')}")
    echo(f"  id       {deployment.get('id')}")
    echo("")
    echo(
        "This tears down the deployment and destroys everything it stores. It "
        "cannot be undone."
    )


def confirm(
    deployment: Dict[str, Any],
    env_name: str,
    *,
    assume_yes: bool,
    interactive: bool,
    echo: Callable[[str], None],
    ask: Callable[[str], bool],
) -> bool:
    """Whether to go ahead. `--yes` answers in advance; nothing else may.

    Without a terminal and without `--yes` this refuses rather than proceeding:
    an unattended run that deletes because nobody was there to object is the
    one behavior this command must not have.
    """
    if assume_yes:
        return True
    if not interactive:
        raise UsageError(
            "deleting a deployment needs confirmation and there is no terminal to "
            "ask on.\n"
            "  Re-run with --yes to confirm in advance. Nothing has been deleted."
        )

    present(deployment, env_name, echo)
    return ask(f"Delete '{deployment.get('name')}'?")


# --------------------------------------------------------------------------
# The request
# --------------------------------------------------------------------------


def request_deletion(api: ApiClient, user_id: int, deployment_id: str) -> None:
    """`DELETE /api/users/{user_id}/deployments/{deployment_id}`.

    Idempotent on the platform: deleting something already being deleted still
    answers 204. A 404 is treated as success for the same reason — the caller
    asked for the deployment to be gone, and it is.
    """
    response = api.delete(f"/api/users/{user_id}/deployments/{deployment_id}")
    if response.status_code == 404:
        return
    if response.status_code == 409:
        message, retryable = describe_conflict(_json_detail(response))
        if retryable:
            message = f"{message}\n  Nothing has been deleted."
        raise FreepodError(message)
    if response.status_code not in (200, 202, 204):
        raise FreepodError(
            f"could not delete deployment {deployment_id}: HTTP {response.status_code} "
            f"{response.text.strip()[:300]}"
        )


# --------------------------------------------------------------------------
# Following the teardown
# --------------------------------------------------------------------------


def wait_until_gone(
    api: ApiClient,
    user_id: int,
    deployment_id: str,
    *,
    timeout: int = ROLLOUT_WAIT_SECONDS,
    poll: float = POLL_SECONDS,
    echo: Callable[[str], None] = _log,
) -> None:
    """Poll until the deployment is `deleted`, or the platform says it failed.

    A teardown that raises leaves the deployment in `error` with `last_error`
    set, exactly as a failed rollout does. Waiting that out to the timeout
    would report "still deleting" for something the platform has already given
    up on, so it is read as the failure it is.
    """
    reported: Optional[str] = None
    deadline = time.monotonic() + timeout

    while True:
        record = read_deployment(api, user_id, deployment_id)
        if record is None:
            return

        status = record.get("status")
        if status == STATUS_DELETED:
            return
        if status == STATUS_ERROR:
            raise FreepodError(
                f"the teardown of deployment {deployment_id} failed on the platform.\n"
                f"  The platform recorded: {record.get('last_error') or 'no error message'}\n"
                f"  The deployment may still hold its hostname. Delete it from "
                f"{api.env.api_base} once the cause is resolved."
            )

        if status != reported:
            echo(f"  {status}...")
            reported = status

        if time.monotonic() >= deadline:
            raise FreepodError(
                f"stopped waiting after {timeout}s. Deployment {deployment_id} is "
                f"still being torn down on the platform — the deletion was not "
                f"canceled, and {PROJECT_FILE} no longer points at it."
            )
        time.sleep(poll)


# --------------------------------------------------------------------------
# The command
# --------------------------------------------------------------------------


def delete(
    api: ApiClient,
    env_name: str,
    *,
    root: Optional[Path] = None,
    assume_yes: bool = False,
    wait: bool = True,
    interactive: bool = True,
    timeout: int = ROLLOUT_WAIT_SECONDS,
    poll: float = POLL_SECONDS,
    echo: Callable[[str], None] = _log,
    ask: Callable[[str], bool] = _ask,
) -> bool:
    """Delete this project's deployment. Returns whether anything was deleted.

    A declined confirmation returns False rather than raising: the user was
    asked and answered, which is not a failure of the command.
    """
    project = require_project(env_name, root)

    if not project.deployment_id:
        raise UsageError(
            f"{project.path} records no deployment, so there is nothing to delete.\n"
            f"  Run `freepod deploy` to create one."
        )

    # `/api/me` first, as everywhere: it is the request that actually exercises
    # the credential, and the deployment read below is scoped by its answer.
    user_id = api.me()["id"]
    deployment_id = project.deployment_id
    record = read_deployment(api, user_id, deployment_id)

    if record is None:
        # Gone already — deleted from the web UI, or from another checkout.
        # Clearing the stale pointer is the whole of what is left to do.
        name = project.deployment_name
        echo(
            f"Deployment {deployment_id}{f' ({name})' if name else ''} no longer "
            f"exists on '{env_name}'; nothing to delete."
        )
        _forget(project, echo)
        return False

    status = record.get("status")
    if status in (STATUS_DELETING, STATUS_DELETED):
        # The platform has already accepted a deletion for this one, so there
        # is nothing to confirm: the destructive decision was made earlier.
        echo(f"Deployment '{record.get('name')}' is already {status}.")
        _forget(project, echo)
        if wait and status == STATUS_DELETING:
            wait_until_gone(
                api, user_id, deployment_id, timeout=timeout, poll=poll, echo=echo
            )
            echo("Deleted.")
        return False

    if not confirm(
        record,
        env_name,
        assume_yes=assume_yes,
        interactive=interactive,
        echo=echo,
        ask=ask,
    ):
        echo("Nothing was deleted.")
        return False

    request_deletion(api, user_id, deployment_id)
    echo(f"Deleting deployment '{record.get('name')}' ({deployment_id})...")

    # Discarded now, not after the wait: the deployment is on its way out and
    # can never serve this project again, so an interrupted wait must not leave
    # the project pointing at it.
    _forget(project, echo)

    if wait:
        wait_until_gone(api, user_id, deployment_id, timeout=timeout, poll=poll, echo=echo)
        echo("Deleted.")
    else:
        echo(
            "Teardown continues on the platform. Its hostname stays claimed until "
            "it finishes."
        )
    return True


def _forget(project: Project, echo: Callable[[str], None]) -> None:
    project.forget_deployment()
    echo(f"Cleared the deployment pointer in {project.path}.")

File: cli-rust/src/delete.rs
//! Deleting the deployment a project points at.
//!
//! Mirrors `delete.py`. The mirror image of `deploy`, and deliberately the only
//! destructive thing this client does. It addresses the deployment recorded in
//! the project file and nothing else.
//!
//! Teardown is asynchronous: `DELETE` answers **204** and moves the deployment
//! to `deleting`; the reconciler uninstalls the release afterwards. The client
//! follows the teardown to `deleted` by default, because the hostname stays
//! claimed until it lands.

use std::path::Path;
use std::time::{Duration, Instant};

use serde_json::Value;

use crate::api::ApiClient;
use crate::deploy::describe_conflict;
use crate::errors::{freepod, usage, Result};
use crate::project::{require_project, Project, PROJECT_FILE};

/// Where a teardown ends.
pub const STATUS_DELETED: &str = "deleted";
/// In flight.
const STATUS_DELETING: &str = "deleting";
const STATUS_ERROR: &str = "error";

/// How often to re-read while following a teardown.
pub const POLL_SECONDS: f64 = 3.0;

/// The deployment, or None if the platform no longer has it.
async fn read_deployment(
    api: &mut ApiClient,
    user_id: u64,
    deployment_id: &str,
) -> Result<Option<Value>> {
    let response = api
        .get(&format!("/api/users/{user_id}/deployments/{deployment_id}"), None)
        .await?;
    if response.status == 404 {
        return Ok(None);
    }
    if !response.is_success() {
        return Err(freepod(format!(
            "could not read deployment {deployment_id}: HTTP {} {}",
            response.status,
            response.text().trim().chars().take(300).collect::<String>()
        )));
    }
    Ok(Some(response.decode()?))
}

/// `https://…` for a deployment's hostname, or None.
fn address(deployment: &Value) -> Option<String> {
    deployment
        .get("hostname")
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .map(|h| format!("https://{h}"))
}

/// Show what is about to be destroyed, by name and by address.
fn present(deployment: &Value, env_name: &str, echo: &dyn Fn(&str)) {
    echo("");
    echo(&format!(
        "About to delete deployment '{}' on '{env_name}':",
        deployment.get("name").and_then(|v| v.as_str()).unwrap_or("?")
    ));
    if let Some(live) = address(deployment) {
        echo(&format!("  address  {live}"));
    }
    echo(&format!(
        "  status   {}",
        deployment.get("status").and_then(|v| v.as_str()).unwrap_or("?")
    ));
    echo(&format!(
        "  id       {}",
        deployment.get("id").and_then(|v| v.as_str()).unwrap_or("?")
    ));
    echo("");
    echo(
        "This tears down the deployment and destroys everything it stores. It \
         cannot be undone.",
    );
}

/// Whether to go ahead. `--yes` answers in advance; nothing else may.
fn confirm(
    deployment: &Value,
    env_name: &str,
    assume_yes: bool,
    interactive: bool,
    echo: &dyn Fn(&str),
    ask: &dyn Fn(&str) -> Result<bool>,
) -> Result<bool> {
    if assume_yes {
        return Ok(true);
    }
    if !interactive {
        return Err(usage(
            "deleting a deployment needs confirmation and there is no terminal to \
             ask on.\n  Re-run with --yes to confirm in advance. Nothing has been \
             deleted.",
        ));
    }
    present(deployment, env_name, echo);
    ask(&format!(
        "Delete '{}'?",
        deployment.get("name").and_then(|v| v.as_str()).unwrap_or("?")
    ))
}

/// `DELETE /api/users/{user_id}/deployments/{deployment_id}`.
async fn request_deletion(
    api: &mut ApiClient,
    user_id: u64,
    deployment_id: &str,
) -> Result<()> {
    let response = api
        .delete(&format!("/api/users/{user_id}/deployments/{deployment_id}"))
        .await?;
    if response.status == 404 {
        return Ok(());
    }
    if response.status == 409 {
        let (message, retryable) = describe_conflict(response.detail().as_deref(), None);
        let message = if retryable {
            format!("{message}\n  Nothing has been deleted.")
        } else {
            message
        };
        return Err(freepod(message));
    }
    if !(response.status == 200 || response.status == 202 || response.status == 204) {
        return Err(freepod(format!(
            "could not delete deployment {deployment_id}: HTTP {} {}",
            response.status,
            response.text().trim().chars().take(300).collect::<String>()
        )));
    }
    Ok(())
}

/// Poll until the deployment is `deleted`, or the platform says it failed.
async fn wait_until_gone(
    api: &mut ApiClient,
    user_id: u64,
    deployment_id: &str,
    timeout: u64,
    poll: f64,
    echo: &dyn Fn(&str),
) -> Result<()> {
    let mut reported: Option<String> = None;
    let deadline = Instant::now() + Duration::from_secs(timeout);

    loop {
        let Some(record) = read_deployment(api, user_id, deployment_id).await? else {
            return Ok(());
        };
        let status = record
            .get("status")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        if status == STATUS_DELETED {
            return Ok(());
        }
        if status == STATUS_ERROR {
            let last_error = record
                .get("last_error")
                .and_then(|v| v.as_str())
                .filter(|s| !s.is_empty())
                .unwrap_or("no error message");
            return Err(freepod(format!(
                "the teardown of deployment {deployment_id} failed on the platform.\n  \
                 The platform recorded: {last_error}\n  The deployment may still hold \
                 its hostname. Delete it from {} once the cause is resolved.",
                api.env.api_base
            )));
        }

        if reported.as_deref() != Some(status.as_str()) {
            echo(&format!("  {status}..."));
            reported = Some(status.clone());
        }

        if Instant::now() >= deadline {
            return Err(freepod(format!(
                "stopped waiting after {timeout}s. Deployment {deployment_id} is still \
                 being torn down on the platform — the deletion was not canceled, and \
                 {PROJECT_FILE} no longer points at it."
            )));
        }
        tokio::time::sleep(Duration::from_secs_f64(poll)).await;
    }
}

/// Delete this project's deployment. Returns whether anything was deleted.
#[allow(clippy::too_many_arguments)] // mirrors the reference signature
pub async fn delete(
    api: &mut ApiClient,
    env_name: &str,
    root: Option<&Path>,
    assume_yes: bool,
    wait: bool,
    interactive: bool,
    timeout: u64,
    poll: f64,
    echo: &dyn Fn(&str),
    ask: &dyn Fn(&str) -> Result<bool>,
) -> Result<bool> {
    let mut project = require_project(root)?;

    if project.env != env_name && project.deployment_id().is_some() {
        return Err(usage(format!(
            "{} records deployment '{}' on '{}', not on '{env_name}'.\n  Re-run without \
             --env (or with --env {}) to delete it.",
            project.path().display(),
            project.deployment_name().unwrap_or(""),
            project.env,
            project.env
        )));
    }

    let Some(deployment_id) = project.deployment_id().map(|s| s.to_string()) else {
        return Err(usage(format!(
            "{} records no deployment, so there is nothing to delete.\n  Run `freepod \
             deploy` to create one.",
            project.path().display()
        )));
    };

    // `/api/me` first, as everywhere.
    let user_id = api
        .me()
        .await?
        .get("id")
        .and_then(|v| v.as_u64())
        .unwrap_or(0);
    let record = read_deployment(api, user_id, &deployment_id).await?;

    let Some(record) = record else {
        // Gone already — clearing the stale pointer is the whole of what is left.
        let name = project.deployment_name().map(|n| format!(" ({n})")).unwrap_or_default();
        echo(&format!(
            "Deployment {deployment_id}{name} no longer exists on '{env_name}'; nothing \
             to delete."
        ));
        forget(&mut project, echo)?;
        return Ok(false);
    };

    let status = record
        .get("status")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    if status == STATUS_DELETING || status == STATUS_DELETED {
        echo(&format!(
            "Deployment '{}' is already {status}.",
            record.get("name").and_then(|v| v.as_str()).unwrap_or("?")
        ));
        forget(&mut project, echo)?;
        if wait && status == STATUS_DELETING {
            wait_until_gone(api, user_id, &deployment_id, timeout, poll, echo).await?;
            echo("Deleted.");
        }
        return Ok(false);
    }

    if !confirm(&record, env_name, assume_yes, interactive, echo, ask)? {
        echo("Nothing was deleted.");
        return Ok(false);
    }

    request_deletion(api, user_id, &deployment_id).await?;
    echo(&format!(
        "Deleting deployment '{}' ({deployment_id})...",
        record.get("name").and_then(|v| v.as_str()).unwrap_or("?")
    ));

    // Discarded now, not after the wait.
    forget(&mut project, echo)?;

    if wait {
        wait_until_gone(api, user_id, &deployment_id, timeout, poll, echo).await?;
        echo("Deleted.");
    } else {
        echo(
            "Teardown continues on the platform. Its hostname stays claimed until it \
             finishes.",
        );
    }
    Ok(true)
}

fn forget(project: &mut Project, echo: &dyn Fn(&str)) -> Result<()> {
    project.forget_deployment()?;
    echo(&format!("Cleared the deployment pointer in {}.", project.path().display()));
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn address_uses_https() {
        let d = json!({ "hostname": "foo.freepod.eu" });
        assert_eq!(address(&d).as_deref(), Some("https://foo.freepod.eu"));
        assert_eq!(address(&json!({})), None);
    }
}
